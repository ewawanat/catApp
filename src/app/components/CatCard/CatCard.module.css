.cat-card-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    background-color: #fff;
    border-radius: 10px;
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
    overflow: hidden;
    width: 700px;
    padding: 20px;
    transition: transform 0.3s ease;
}

.cat-card-container:hover {
    transform: translateY(-5px);
}

.image-fit {
    object-fit: contain;
    /* Ensure the image maintains its aspect ratio */
    width: 100%;
    /* Make the image fill the container */
    height: 100%;
    /* Make the image fill the container */
    display: block;
    /* Remove any unwanted space below the image */
}

.actions-container {
    display: flex;
    justify-content: space-around;
    align-items: center;
    width: 100%;
    margin-top: 10px;
}

.actions-container .MuiIconButton-root {
    background-color: #f0f0f0;
    border-radius: 50%;
    padding: 8px;
    transition: background-color 0.3s ease;
}

.actions-container .MuiIconButton-root:hover {
    background-color: #e0e0e0;
}

.actions-container .MuiIconButton-root:active {
    background-color: #d0d0d0;
}

.votes-container {
    margin-top: 15px;
    display: flex;
    flex-direction: column;
    align-items: center;
    font-family: 'Roboto', sans-serif;
    color: #333;
}

.votes-container Typography {
    font-size: 1.2rem;
    font-weight: bold;
    color: #2c3e50;
}

.votes-container .MuiTypography-root {
    font-size: 1.2rem;
    font-weight: 600;
}

.votes-container .MuiTypography-body2 {
    color: #3498db;
    /* Add color to the text */
}

@media (max-width: 600px) {
    .cat-card-container {
        width: 90%;
        padding: 15px;
    }

    .image-fit {
        max-height: 150px;
    }
}