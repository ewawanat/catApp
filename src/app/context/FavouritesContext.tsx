"use client";
import React, { createContext, useContext, useState } from "react";
import { Cat } from "../page";

type FavouritesContextType = {
    favourites: Cat[];
    addFavourite: (cat: Cat) => void;
    removeFavourite: (catId: string) => void;
};

const FavouritesContext = createContext<FavouritesContextType | undefined>(undefined);

export const FavouritesProvider = ({ children }: { children: React.ReactNode }) => {
    const [favourites, setFavourites] = useState<Cat[]>([]);

    const addFavourite = (cat: Cat) => {
        setFavourites((prev) => [...prev, cat]);
    };

    const removeFavourite = (catId: string) => {
        setFavourites((prev) => prev.filter((cat) => cat.id !== catId));
    };

    return (
        <FavouritesContext.Provider value={{ favourites, addFavourite, removeFavourite }}>
            {children}
        </FavouritesContext.Provider>
    );
};

export const useFavourites = () => {
    const context = useContext(FavouritesContext);
    if (!context) {
        throw new Error("useFavourites must be used within a FavouritesProvider");
    }
    return context;
};
