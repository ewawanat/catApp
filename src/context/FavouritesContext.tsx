"use client";
import React, { createContext, useContext, useState } from "react";
import { Cat } from "@/app/page";

type FavouritesContextType = {
    favourites: Cat[];
    addFavourite: (cat: Cat) => void;
    removeFavourite: (catId: string) => void;
};

const FavouritesContext = createContext<FavouritesContextType | undefined>(undefined);

export const FavouritesProvider = ({ children }: { children: React.ReactNode }) => {
    const [favourites, setFavourites] = useState<Cat[]>([]);

    const addFavourite = (cat: Cat) => {
        setFavourites((prev) => [...prev, cat]);
        //add logic for making the POST request to update the favourites on backend
    };

    const removeFavourite = (catId: string) => {
        setFavourites((prev) => prev.filter((cat) => cat.id !== catId));
        //add logic for making the POST request to update the favourites on backend

    };

    return (
        <FavouritesContext.Provider value={{ favourites, addFavourite, removeFavourite }}>
            {children}
        </FavouritesContext.Provider>
    );
};

export const useFavourites = () => {
    const context = useContext(FavouritesContext);
    if (!context) {
        throw new Error("useFavourites must be used within a FavouritesProvider");
    }
    return context;
};