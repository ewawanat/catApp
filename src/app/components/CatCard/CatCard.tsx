"use client";
import { useState } from "react";
import Link from "next/link";
import Image from "next/image";
import IconButton from "@mui/material/IconButton";
import FavoriteIcon from "@mui/icons-material/Favorite";
import FavoriteBorderIcon from "@mui/icons-material/FavoriteBorder";
import ThumbUpIcon from "@mui/icons-material/ThumbUp";
import ThumbDownIcon from "@mui/icons-material/ThumbDown";
import Typography from "@mui/material/Typography";

import { Cat } from "@/app/page";
import { useFavourites } from "../../../context/FavouritesContext";
import styles from "./CatCard.module.css";

type CatProps = {
    cat: Cat;
};

const CatCard = ({ cat }: CatProps) => {
    const { favourites, addFavourite, removeFavourite } = useFavourites();
    const isFavourite = favourites.some((fav) => fav.id === cat.id);
    const [votes, setVotes] = useState(cat.votes || 0);

    const toggleFavorite = () => {
        if (isFavourite) {
            removeFavourite(cat.id);
        } else {
            addFavourite(cat);
        }
    };

    const voteForCat = async (value: number) => {
        try {
            const response = await fetch("https://api.thecatapi.com/v1/votes", {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    "x-api-key": process.env.NEXT_PUBLIC_CAT_API_KEY || "",
                },
                body: JSON.stringify({
                    image_id: cat.id,
                    sub_id: "user-123", // Hardcoded user ID
                    value,
                }),
            });

            if (!response.ok) {
                throw new Error("Failed to vote for cat");
            }
        } catch (error) {
            console.error("Error voting for cat:", error);
        }
    };

    const handleVoteUp = async () => {
        setVotes((prevVotes) => prevVotes + 1); // Optimistically update the state
        await voteForCat(1); // Send POST request for upvote
    };

    const handleVoteDown = async () => {
        setVotes((prevVotes) => prevVotes - 1); // Optimistically update the state
        await voteForCat(-1); // Send POST request for downvote
    };

    return (
        <div className={styles["cat-card-container"]}>
            <Link
                href={{
                    pathname: `/breeds/${cat.id}`,
                    query: {
                        width: cat.width,
                        height: cat.height,
                        image: cat.url,
                    },
                }}
            >
                <Image
                    src={cat.url}
                    alt={`Cat ${cat.id}`}
                    width={cat.width}
                    height={cat.height}
                    className={styles["image-fit"]}
                />
            </Link>
            <div className={styles["actions-container"]}>
                <IconButton onClick={toggleFavorite} aria-label={isFavourite ? "Remove from favourites" : "Add to favourites"}>
                    {isFavourite ? <FavoriteIcon color="error" /> : <FavoriteBorderIcon />}
                </IconButton>
                <IconButton onClick={handleVoteUp}>
                    <ThumbUpIcon />
                    <Typography variant="body2" ml={1}>Vote Up</Typography>
                </IconButton>
                <IconButton onClick={handleVoteDown}>
                    <ThumbDownIcon />
                    <Typography variant="body2" ml={1}>Vote Down</Typography>
                </IconButton>
            </div>
            <div className={styles["votes-container"]}>
                <Typography variant="body2">Votes: {votes}</Typography>
            </div>
        </div>
    );
};

export default CatCard;
