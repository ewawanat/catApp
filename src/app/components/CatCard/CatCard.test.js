import * as React from 'react'
import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import CatCard from './CatCard';
import { useFavourites } from '../../../context/FavouritesContext';

jest.mock('../../../context/FavouritesContext', () => ({
    useFavourites: jest.fn(),
}));
expect.extend(toHaveNoViolations); // Extending jest to use axe matcher


describe('CatCard', () => {
    const cat = {
        id: '1',
        url: 'https://example.com/cat.jpg',
        width: 500,
        height: 500,
        votes: 10,
    };
    it('renders and allows voting', async () => {
        // Mock the return value of useFavourites
        useFavourites.mockReturnValue({
            favourites: [{ id: '1', url: 'https://example.com/cat.jpg', width: 500, height: 500, votes: 10 }],
            addFavourite: jest.fn(),
            removeFavourite: jest.fn(),
        });

        render(<CatCard cat={cat} />);
        // Check if the image is rendered
        const image = screen.getByRole('img', { name: /cat 1/i });
        expect(image).toBeInTheDocument();

        // Check if the votes are displayed correctly
        const votes = screen.getByText(/Votes:/);
        expect(votes).toHaveTextContent('Votes: 10');

        // Simulate clicking the vote up button
        const voteUpButton = screen.getByText('Vote Up');
        await userEvent.click(voteUpButton);

        // Check if the votes count increases
        expect(screen.getByText(/Votes:/)).toHaveTextContent('Votes: 11');
    });

    it('toggles the favourite status', async () => {
        const addFavouriteMock = jest.fn();
        const removeFavouriteMock = jest.fn();

        // Mock the return value for the favourites context
        useFavourites.mockReturnValue({
            favourites: [{ id: '1', url: 'https://example.com/cat.jpg', width: 500, height: 500, votes: 10 }],
            addFavourite: addFavouriteMock,
            removeFavourite: removeFavouriteMock,
        });



        render(<CatCard cat={cat} />);

        // Check if the favourite button is rendered correctly
        const favouriteButton = screen.getByRole('button', { name: /remove from favourites/i });
        await userEvent.click(favouriteButton);

        // Check if the removeFavourite function was called
        expect(removeFavouriteMock).toHaveBeenCalledWith(cat.id);

        // Update mock to simulate the cat not being a favourite anymore
        useFavourites.mockReturnValue({
            favourites: [],
            addFavourite: addFavouriteMock,
            removeFavourite: removeFavouriteMock,
        });

        // Re-render the component to reflect the updated favourites state
        render(<CatCard cat={cat} />);

        const addFavouriteButton = screen.getByRole('button', { name: /add to favourites/i });
        await userEvent.click(addFavouriteButton);

        // Check if the addFavourite function was called
        expect(addFavouriteMock).toHaveBeenCalledWith(cat);
    });
    it('matches the snapshot', () => {
        const addFavouriteMock = jest.fn();
        const removeFavouriteMock = jest.fn();

        // Mock the return value for the favourites context
        useFavourites.mockReturnValue({
            favourites: [{ id: '1', url: 'https://example.com/cat.jpg', width: 500, height: 500, votes: 10 }],
            addFavourite: addFavouriteMock,
            removeFavourite: removeFavouriteMock,
        });
        const { asFragment } = render(<CatCard cat={cat} />);

        // Take a snapshot of the component
        expect(asFragment()).toMatchSnapshot();
    });
    it('should have no accessibility violations', async () => {
        const { container } = render(<CatCard cat={cat} />);

        // Check for accessibility violations
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
});