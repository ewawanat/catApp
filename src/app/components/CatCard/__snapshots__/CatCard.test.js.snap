// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatCard matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="cat-card-container"
  >
    <a
      href="/breeds/1?width=500&height=500&image=https%3A%2F%2Fexample.com%2Fcat.jpg"
    >
      <img
        alt="Cat 1"
        class="image-fit"
        data-nimg="1"
        decoding="async"
        height="500"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fexample.com%2Fcat.jpg&w=1080&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Fcat.jpg&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com%2Fcat.jpg&w=1080&q=75 2x"
        style="color: transparent;"
        width="500"
      />
    </a>
    <div
      class="actions-container"
    >
      <button
        aria-label="Remove from favourites"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium css-n4bcw2-MuiSvgIcon-root"
          data-testid="FavoriteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54z"
          />
        </svg>
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ThumbUpIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M1 21h4V9H1zm22-11c0-1.1-.9-2-2-2h-6.31l.95-4.57.03-.32c0-.41-.17-.79-.44-1.06L14.17 1 7.59 7.59C7.22 7.95 7 8.45 7 9v10c0 1.1.9 2 2 2h9c.83 0 1.54-.5 1.84-1.22l3.02-7.05c.09-.23.14-.47.14-.73z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body2 css-1hpvxul-MuiTypography-root"
        >
          Vote Up
        </p>
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ThumbDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15 3H6c-.83 0-1.54.5-1.84 1.22l-3.02 7.05c-.09.23-.14.47-.14.73v2c0 1.1.9 2 2 2h6.31l-.95 4.57-.03.32c0 .41.17.79.44 1.06L9.83 23l6.59-6.59c.36-.36.58-.86.58-1.41V5c0-1.1-.9-2-2-2m4 0v12h4V3z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body2 css-1hpvxul-MuiTypography-root"
        >
          Vote Down
        </p>
      </button>
    </div>
    <div
      class="votes-container"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
      >
        Votes: 10
      </p>
    </div>
  </div>
</DocumentFragment>
`;
